nixos/modules/services/network-filesystems/tahoe/utils.nix

```diff
@@ -0,0 +1,209 @@
+    (optionalAttrs (cfg ? client.shares.happy)  { client = { "shares.happy"    = cfg.client.shares.happy; }; })
+    (optionalAttrs (cfg ? client.shares.total)  { client = { "shares.total"    = cfg.client.shares.total; }; })
+
+    (optionalAttrs (cfg ?  helper.enable)       { helper = { "enabled"         = cfg.helper.enable; }; })
+
+    (optionalAttrs (cfg ? sftpd.enable)             { sftpd =  { "enabled"           = cfg.sftpd.enable; }; })
+    (optionalAttrs (cfg ? sftpd.port)               { sftpd =  { "port"              = "tcp:${toString cfg.sftpd.port}"; }; })
+    (optionalAttrs (cfg ? sftpd.hostPublicKeyFile)  { sftpd =  { "host_pubkey_file"  = cfg.sftpd.hostPublicKeyFile; }; })
+    (optionalAttrs (cfg ? sftpd.hostPrivateKeyFile) { sftpd =  { "host_privkey_file" = cfg.sftpd.hostPrivateKeyFile; }; })
+    (optionalAttrs (cfg ? sftpd.accounts.file)      { sftpd =  { "accounts.file"     = cfg.sftpd.accounts.file; }; })
+  ];
+
+  # Generate the ini-format Tahoe-LAFS `tahoe.cfg` configuration file from the
+  # NixOS service configuration value.
+  #
+  # The attrset argument is the top of a single node's configuration - for
+  # example, `config.services.nodes.bar_node`.
+  #
+  # The result is the path of the generated configuration file.
+  #
+  # str -> str -> attrset -> Path
+  mkConfig = kind: name: cfg:
+    let
+      mergePieces = foldl recursiveUpdate {};
+
+      configPieces =
+        if kind == "node" then nodeConfigPieces
+        else if kind == "introducer" then introducerConfigPieces
+        else throw "Tahoe-LAFS kind must be node or introducer, not ${kind}";
+
+      configPathFragment = "tahoe-lafs/${kind}-${name}.cfg";
+      configValue = mergePieces ([cfg.settings] ++ configPieces cfg);
+      configPath = settingsFormat.generate configPathFragment configValue;
+    in
+      configPath;
+
+  # Construct the NixOS systemd service configuration for all of the given
+  # node configurations.
+  #
+  # The result is used as the value of `systemd.services`.
+  #
+  # str -> attrset -> attrset
+  mkServices = kind: nodes: forEach nodes (name: cfg:
+    nameValuePair "tahoe.${kind}-${name}" {
+      description = "Tahoe LAFS node ${name}";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = mkServiceConfig cfg.package kind name (mkConfig kind name cfg);
+    }
+  );
+
+  # Construct the NixOS static user configuration for all of the given node
+  # configurations.
+  #
+  # The result is used as the value of `users.users`.
+  #
+  # str -> str -> attrset
+  mkUsers = kind: nodes: forEach nodes (name: cfg:
+    nameValuePair (userName kind name) {
+      description = "Tahoe node user for ${kind} ${name}";
+      isSystemUser = true;
+      group = groupName kind name;
+    }
+  );
+
+  # Like mkUsers, but for group configuration.
+  mkGroups = kind: nodes: forEach nodes (name: cfg:
+    nameValuePair (groupName kind name) { }
+  );
```

Comment:
systemd DynamicUsers?  We could, and I would like to - ideally as a follow-up.  I currently do not know how to manage the state upgrade that this will require (the service currently places data in /var/db; as I understand it, `DynamicUser` will require the state to be in `/var/lib` instead).

Reply:
Alright. Migrating state could be complicated. I guess a privileged ExecStartPre would be the right way